packages/yarnpkg-fslib/sources/path.ts

```diff
@@ -6,15 +6,40 @@ enum PathType {
   Native,
 }
 
-export type PortablePath = string & { _path_type: PathType.File | PathType.Portable };
-export type NativePath = string & { _path_type?: PathType.File | PathType.Native };
+enum PathRelativity {
+  Absolute,
+  Relative,
+}
+
+export type AbsolutePortablePath = string & {
+  _path_type: PathType.File | PathType.Portable;
+  _path_relativity: PathRelativity.Absolute;
+};
+export type AbsoluteNativePath = string & {
+  _path_type?: PathType.File | PathType.Native;
+  _path_relativity: PathRelativity.Absolute;
+};
+export type RelativePortablePath = string & {
+  _path_type: PathType.File | PathType.Portable;
+  _path_relativity: PathRelativity.Relative;
+};
+export type RelativeNativePath = string & {
+  _path_type?: PathType.File | PathType.Native;
+  _path_relativity: PathRelativity.Relative;
+};
```

Comment:
Do you think it would make sense to turn that into "bitfields"? For example, instead of `RelativePortablePath`, we'd have `Path<Portable | Relative>` (and `Path<Portable>` would accept both relative and absolute paths). Might be more scalable, if we want more types of paths? 🤔

Reply:
That makes sense, but I'll have to toy around and see what's possible, because

```ts
declare let relativePortablePath: Path<Portable | Relative>`;
declare let portablePath: Path<Portable>;

// we need this to work
portablePath = relativePortablePath;
// and this to not work
relativePortablePath = portablePath
```

With type unions you'd usually expect the opposite to happen.

Reply:
This could work, see [Playground Link](https://www.typescriptlang.org/play?ts=3.9.2#code/KYOwrgtgBAYgNgQwOYGcoG8BQUoAUD2ATgC4IBGcwANNlAHILECWAbtZrQErCLNs04AgmRT44YYuwC+HYgE8ADsDxFSFZQF5YiVADoCJcpQDcmeUvqNWm7chS6GfYKfPKYTSlC3w7+1UeUAH1s9R2tTM0Vlbl5rLxD7GKs2FyioYVFxSXifPQyxCWcOYAAPBVUMWgM1SgFLJzr3Wo4cJIbafKzpDgATYABjREJlfvwQFGIoAGsAFSiALigwECYARzBlFDkIMjEI1ygASRRBFBQmJBAAmfwAHlPzy+BqdLOLkGAeqiOTt8uA77HEDEB7vAIAPniAG1QU8egBdKClSQgHpoGF-D7AREAfh+sKu6igiyBIMxAVSFkOADMAKKre7fABCgJQ9LACDgrOB7M5kK0xwJ1zugmZrKF6hutxZ6VZvK5Px56z53OI8vB+zSuEYAAtbjAkSUUWiEvyoBNCEwQEgoAAyI50hkwb5hfiVHBQ2ZReE4xYACmpOhQixgAEovJCPmxCKYpN8sB6vUp4f7A3YQ+GNJHgNHYxrimUKgdXcBtcQdfEy3qS-nSuUSFADtUAlXK7rbs31LXCw2Dk1QAgIDYq-qPEEoG1rPnMH1BghhlBKJMQApFiWq6ZZ0NlEuoApVypDOoNxwAPSnqAAZR1+DAcB6UF25ag1IQHkwK-i+9M+-iK4iW7zjuwCTCgB4WlaSAROeV43neD4AO5EFMH4KPEYGmDB163veIAAOSTEhhAob+WgYWeF7YfBj74M+REoWBf4KKYjFaN+vQDNui4gS+IZjlcQ7QRe-YCcAaAoHB96PsoCDkkSxD4Hu-hEggqJQFcTh7rqKCofE1I-mhWj6ZgrEvkJj4SOptFQCwTD9Mo0YoAgmDUkxpiuWxzEuehXkzpxQHccuCAHiOJZQMEnSFPmgELruICECF7ZhcEk5sNOMHDLEbBQEQUDkJkhRQD0YwEVAECMJIhBQAhOqgHlmKQY2imhckwDgrpWhXF5n6dQl5lkJZ-SqaVsmPCAeBJa1kIKXlUCZa1OVVflBTZAouofsFbkfglW1+XOsU8WtiXlh2ymeBFIgrW1m7+QdkwKAliwjp250TjwU3mfNmm5ctXRFSVkzlcQlXVbV42je8jUzc9Z1tZgpFaV5CMPb5MEDZMQ34ZMEOXBNJ0vW1TWzV9cQ-Zdf1reW8ObZ5P47bTFFQDMOpMGg9EoN86P-WJWNlQgUybKAKDAHtXG7vOx16gT4XpOThUpe9TgakAA)

There's one big caveat: nothing stops you from writing `Path<Absolute | Relative>`, which doesn't make any sense because a path can't be both absolute and relative.

We could block this, but it complicates the `Path` type a little bit, see [Playground Link](https://www.typescriptlang.org/play?ts=3.9.2#code/KYOwrgtgBAYgNgQwOYGcoG8BQUoAUD2ATgC4IBGcwANNlAHILECWAbtZrQErCLNs04AgmRT44YYuwC+HYgE8ADsDxFSFZQF5YiVADoCJcpQDcmeUvqNWm7chS6GfYKfPKYTSlC3w7+1UeUAH1s9R2tTM0Vlbl5rLxD7GKs2FyioYVFxSXifPQyxCWcOYAAPBVUMWgM1SgFLJzr3Wo4cJIbafKzpDgATYABjREJlfvwQFGIoAGsAFSiALigwECYARzBlFDkIMjEI1ygASRRBFBQmJBAAmfwAHlPzy+BqdLOLkGAeqiOTt8uA77HEDEB7vAIAPniAG1QU8egBdKClSQgHpoGF-D7AREAfh+sKu6igiyBIMxAVSFkOADMAKKre7fABCgJQ9LACDgrOB7M5kK0xwJ1zugmZrKF6hutxZ6VZvK5Px56z53OI8vB+zSAH0AOpMYgAC1wjANtxgSJKKLRCX5RzpDJg3zC-EqOChsyi8JxiwAFNSdChFjAAJReSEfNiEUxSb5YN0epTw33+uxB0MacPASPRjWYUrlEhQA7Gw1mi1WtC5FC2wXkyV3R3pEQFbLBNrWb4R4CEb4TQhMEBIKAAMiguv1RpNZvBubzZQqB2dwBLBviK9uS9z+YXaWqARXa6ne-UW-nhYOTVACAgNnXl6gbZ4yWAs76gwQwyglEmIAUiyXK6mG+QzKN+UAKH+KiGOogEcAA9HBUAAMoGvgYBwD0UC7IaUDUggHiYL+8QQaYEHxL+ETAR+oHAJMKCQX2A5IBECHIah6GYQA7kQUyEQo8T0aYrEoWhGEgAA5JM3GELxZFaIJ8GISJHFYfgOHSbx9HkQophaVoJG9AMIFfrRuFBh4V43ixiGXlcN5oCg7EYVhygIHWnjEPg4H+ESCColAVxOOBJooHx8TUqR-FaBFmB6bh1lYRIAVqVALBMP0yiRigCCYNS2mmHl+k6blAnFZgVGfmBVyQeuS4Pk2mSFLmFU0T+hA1VOdWPrEbCzqxww9coRBQOQjXZD0YySVAECMJIhBQJxBqgCNmJMUWXm1c+4JhVo1WmERu3tQlZBJf0flTW5jwgHgnVbetI1QANz5QMNo0tsoCgmoRCBRQFxUgO1+WGe+lWmZ9HWlsenjBJ0TVAUZ1EmZMCjtYs65Q0EUDtr1CVPUFr3Nl0UATRJkwzcQc0LUt12Xe8a2eTdkM+ZQ21ycFxVsyjZWsSdkxnaTK1XYzpoY5CDMII9T74-Nb1E59hqYODxEc4DRUJTMBpMGgBxawFWbduVCOg5MH4QyLzOY7DrZY1L1gakAA).

Reply:
What about using `&` instead of `|`? Cf [Playground Link](https://www.typescriptlang.org/play?ts=3.9.2#code/C4TwDgpgBACg9gJ2AQwEYBtoF4oG8D6+AZgM7oCWq+YiKGEAXFMAgK4QC+A3AFCiRQAcsmDkAbtjyFSFKjSRpMAfiZFk6Ep159w0AEoR0I8ZILEylfAkPGJTFu246BAQVQk46VsFPSLVayNROyg1DS0eZ2gAMXJMKBwzGUt5OkwmZAA7ECd+aBgRAAsAIWRNBKlzWWpaRQgVKFQ4TwgsgBooP2rA23qmJpaspx4IAA9UvB4oWFr6NqmhXvnp2Mx5hYMgk2WoNw8vH3mOSIATCABjI2soc7hMkmAoAGsAFV0mVkzyAEd2KBIQABbJrobR5WBFAA80SgYx8mROJAhwBKZQgAD4Kg8EORMgBzKAAMig0W0Ywm4OEwQgBRRFVphUhVJM6N45NozF0MwU9AZ9Kh8B5mFZI3GHPBqwgmWQgMkDOhcWgxM2vRFPDOl2Q10wj0yYCYzIkDN4Gqu0B1UDA+u5aRpRW0AHoHVAAMqFOCsdAnRpwOlqOI8PUVK28K0VPXaU1a80QR4ka3Y3F4x3Ot0er1QADuiCegbAWLAvCdrvdnpOmQA5I9swhc2GcPGUyX096mnSa7n4+HCzwuzgQ6cLmaoBaiExJdLZU2JzKIEiSKWM6hoGUSOQ8dL6Mw4JbZvEst7pdTLUUSHmKkRQ-mcJfe9fQk3UN4oJlfVAxORztAJAgSMgeEQ3a8IB-Y9iB-w9uqQ7RiOsYvsg1ryoaSq7O4njeBiJrQdqcGZAgiFQshRJQCq1JqsWPTHogUBoPsGFQCcdxVlAgIiD4CBZoUUo0SQa4bkm27IoyyHoueODSj2Qbifhj7PucWTMau66ZEJTKqoJyBQJRJhQNRtHoT4J4ooGCFAYG+FmVBmo4Y8YAIUw8qCraxF7AZmFWcOFpgPhDkCnuKGkSyTbaRIukcfpBzQIxlaPKxwDsZx3FKfx+KCY5-mifWJ49ll3mQcWT6PPJMU8XxKnpUKGIaVpNhUeFaGRUZhQ8HZ94Dnlwb5c6LyFOQSIdiQHSFQxcBziVrFPNAmj3BAHkwRaWoESikJOXULkNfRyq1UFPBAA).

Reply:
Oh I missed that 🤦 yeah that makes things a lot simpler, let me try that out :)

Reply:
Okay, I've toyed around with it and it's a lot cleaner like this!
It doesn't work out of the box, we'd need to add these flags on an extra property on the `Path` type because otherwise typescript gets confused, i.e.

```ts
// setup:
type Native = {_path_type?: PathType.File | PathType.Native};
type Portable = {_path_type: PathType.File | PathType.Portable};

type Flags = Native | Portable;
type Path<F extends Flags> = string & F;

type FSPath<P extends Path> = P | number;

// problem:

class ProxiedFS<P extends Path, IP extends Path> {
  private fsMapToBase(p: FSPath<P>) {
    // `typeof p` == all possible types here, while we expect 'number'|'string'
    if (typeof p === `number`) {
      return p; // <-- typeof p yields string|number now? :confused: 
    } else {
      return this.mapToBase(p); // strangely typescript does know p is a string here :trollface: 
    }
  }
}
```

Because of the blanket `& F` in the `Path` type definition typescript seems to forget it is always a string. That actually makes sense though, nothing is stopping someone from using

```ts
type Troll = Map<string, unknown> & unique symbol & Relative;

declare const trollPath: Path<Troll>;
// so trollPath is both a string, a Map, and a symbol
// -> no wonder typescript is confused in the code example above
```

Things turn back to normal if we change the `Path` interface to

```ts
type Path<F extends Flags> = string & {_flags: F};
```

This needs a bit of extra type fu to allow strings to be assigned to `Path<Native>` but it's not too complex.